src/NuGet.Core/NuGet.Configuration/Settings/SettingFactory.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright(c) .NET Foundation.All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml.Linq;
+
+namespace NuGet.Configuration
+{
+    internal static class SettingFactory
+    {
+        internal static SettingBase Parse(XNode node, SettingsFile origin)
+        {
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            if (node is XText textNode)
+            {
+                return new SettingText(textNode, origin);
+            }
+
+            if (node is XElement element)
+            {
+                if (string.Equals(element.Parent?.Name.LocalName, ConfigurationConstants.Configuration, StringComparison.OrdinalIgnoreCase))
+                {
+                    return new ParsedSettingSection(element, origin);
+                }
+                else if (string.Equals(element.Name.LocalName, ConfigurationConstants.Add, StringComparison.OrdinalIgnoreCase))
+                {
+                    if (string.Equals(element.Parent?.Name.LocalName, ConfigurationConstants.PackageSources, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return new SourceItem(element, origin);
+                    }
+
+                    return new AddItem(element, origin);
+                }
+                else if (string.Equals(element.Name.LocalName, ConfigurationConstants.Clear, StringComparison.OrdinalIgnoreCase))
+                {
+                    return new ClearItem(element, origin);
+                }
+                else if (string.Equals(element.Parent?.Name.LocalName, ConfigurationConstants.CredentialsSectionName, StringComparison.OrdinalIgnoreCase))
+                {
+                    return new CredentialsItem(element, origin);
+                }
+
+                return new UnknownItem(element, origin);
+            }
+
+            return null;
+        }
+
+        internal static IEnumerable<T> ParseChildren<T>(XElement xElement, SettingsFile origin, bool canBeCleared) where T : SettingElement
+        {
+            var children = new List<T>();
+
+            var descendants = xElement.Elements().Select(d => Parse(d, origin)).OfType<T>().Distinct();
```

Comment:
It will dedup using the `Equals` function. This is intended because two items that match in `Equals` are referring to the same setting, so we only want to take the first and ignore the rest. This is the previous behavior of settings and this is what `Disctict()` does here.